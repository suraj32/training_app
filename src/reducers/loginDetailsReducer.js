import { LOGIN_REDUCER } from "../shared/actionConstants";

const initialState = {
  email: "",
  password: "",
  emailErrorText: "",
  passwordErrorText: "",
  userDetails: {},
};

const loginDetailsReducer = (state = initialState, action) => {
  switch (action.type) {
    case LOGIN_REDUCER.SET_EMAIL:
      return { ...state, email: action.value };
    case LOGIN_REDUCER.SET_PASSWORD:
      return { ...state, password: action.value };
    case LOGIN_REDUCER.SET_EMAIL_ERROR:
      return { ...state, emailErrorText: action.value };
    case LOGIN_REDUCER.SET_PASSWORD_ERROR:
      return { ...state, passwordErrorText: action.value };
    case LOGIN_REDUCER.SET_USER_DETAILS:
      return { ...state, userDetails: action.value };
    default:
      return state;
  }
};

export default loginDetailsReducer;
