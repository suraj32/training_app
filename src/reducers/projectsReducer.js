import { PROJECT_REDUCER } from "../shared/actionConstants";

const initialState = {
  projects: [],
};

const projectsReducer = (state = initialState, action) => {
  switch (action.type) {
    case PROJECT_REDUCER.SET_PROJECTS:
      return { ...state, projects: action.value };
    default:
      return state;
  }
};

export default projectsReducer;
