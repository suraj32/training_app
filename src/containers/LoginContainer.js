import React, { useReducer } from "react";
import LoginComponent from "../components/LoginComponent";
import * as yup from 'yup';
import apiHelper from "../apis/apiHelper";
import ProductList from "../components/ProductList"
import loginDetailsReducer from "../reducers/loginDetailsReducer"
import { LOGIN_REDUCER } from "../shared/actionConstants";

const LoginContainer = () => {
  const initialState = {
    email: "",
    password: "",
    emailErrorText: "",
    passwordErrorText: "",
    userDetails: {},
  };

  const [loginDetails, dispatch] = useReducer(loginDetailsReducer, initialState);

  const { email, password, userDetails  } = loginDetails

  let schema = yup.object().shape({
    email: yup.string().email().required(),
    password: yup.string().min(6).required(),
  });

  const validateData = () => {
    schema.validate({ email: email, password: password }, { abortEarly: false })
    .then(() => {
      apiHelper('post', 'https://api.taiga.io/api/v1/auth',
        {username: email, password: password, type: 'normal'}).then((data) => {
        dispatch({ type: LOGIN_REDUCER.SET_USER_DETAILS, value: data });
        console.log(userDetails)
      })
    })
    .catch((err) => {
      err.inner.forEach((ele) => {
        dispatch({ type: `SET_${ele.path.toUpperCase()}_ERROR`, value: ele.message });
      });
    });
  };

  return (
    <LoginComponent loginDetails={loginDetails} dispatch={dispatch} validateData={validateData} />
  );
}

export default LoginContainer;
