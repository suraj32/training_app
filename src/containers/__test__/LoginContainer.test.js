import { render, fireEvent, waitFor, screen, waitForElement } from '@testing-library/react'
import LoginContainer from '../LoginContainer'
import { Provider } from 'react-redux'
import { createStore, combineReducers } from 'redux'
import { loginDetailsReducer } from '../../reducers/loginDetailsReducer'
import projectsReducer from '../../reducers/projectsReducer'
import React from 'react';

describe("LoginContainer", () => {
  it("Must return Login component", () => {
    const { asFragment } = render(
      <Provider store={createStore(combineReducers({loginDetailsReducer, projectsReducer}))}>
        <LoginContainer />
      </Provider>
    )
    console.log(asFragment())
    expect(asFragment()).toMatchSnapshot();
  })

  it("Must validate email/password field", async () => {
    const { queryByText } = render(
      <Provider store={createStore(combineReducers({loginDetailsReducer, projectsReducer}))}>
        <LoginContainer />
      </Provider>
    )

    // fireEvent.click(queryByText("Submit"))
    // let emailError = await waitForElement(() => queryByText('email is a required field'))
    // expect(emailError).toBeDefined();
    // const passwordError = await waitForElement(() => queryByText('password is a required field'))
    // expect(passwordError).toBeDefined();

    fireEvent.change(screen.getByLabelText(/email/i), { taregt: { value: 'test' } })
    fireEvent.click(queryByText("Submit"))
    emailError = await waitForElement(() => queryByText('email must be a valid email'))

    expect(asFragment()).toMatchSnapshot();
    expect(emailError).toBeDefined();
  })
})
