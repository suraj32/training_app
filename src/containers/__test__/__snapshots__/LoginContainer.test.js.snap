// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginContainer Must return Login component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="align-items-center row"
    >
      <div
        class="mt-5 col-sm-6 offset-sm-3"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="exampleEmail"
                >
                  Email
                </label>
                <input
                  aria-invalid="false"
                  class="form-control"
                  id="exampleEmail"
                  name="email"
                  placeholder="with a placeholder"
                  type="email"
                  value=""
                />
                <div
                  class="invalid-feedback"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="examplePassword"
                >
                  Password
                </label>
                <input
                  aria-invalid="false"
                  class="form-control"
                  id="examplePassword"
                  name="password"
                  placeholder="password placeholder"
                  type="password"
                  value=""
                />
                <div
                  class="invalid-feedback"
                />
              </div>
              <button
                class="btn btn-secondary"
                type="button"
              >
                Submit
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
